kernel/src/process/clone.rs

```diff
@@ -693,25 +694,91 @@ fn create_child_process(
     child_proc
 }
 
-fn set_parent_and_group(parent: &Process, child: &Arc<Process>) {
-    // Lock order: children of process -> process table -> group of process
-    // -> group inner -> session inner
-    let mut children_mut = parent.children().lock();
+fn set_parent_and_group(clone_flags: CloneFlags, parent: &Arc<Process>, child: &Arc<Process>) {
+    loop {
+        let real_parent = clone_parent(clone_flags, parent);
+
+        // Lock the parent's children before checking its status.
+        let mut children_mut = real_parent.children().lock();
+
+        if children_mut.is_none() {
+            // The real parent is concurrently exiting group.
+            // Wait until it finishes exiting and then retry.
```

Comment:
So, this is not a "wait until" case. The next loop should guarantee a new parent and new progress.